devices/ICub/src/ICub.cpp

```diff
@@ -0,0 +1,388 @@
+
+    const WearableName wearableName = "ICub" + wearable::Separator;
+
+    class wrenchPort : public yarp::os::BufferedPort<yarp::os::Bottle> {
+    public:
+        std::vector<double> wrenchValues;
+        bool firstDataReceived = false;
+        std::mutex mtx;
+
+        using yarp::os::BufferedPort<yarp::os::Bottle>::onRead;
+        virtual void onRead(yarp::os::Bottle& wrench) {
+            if (!wrench.isNull()) {
+                std::lock_guard<std::mutex> lock (mtx);
+
+                this->wrenchValues.clear(); // Clear the vector of previous values
+
+                if (this->wrenchValues.empty()) {
+                    this->wrenchValues.resize(wrench.size());
+                    for (size_t i = 0; i < wrench.size(); ++i) {
+                        this->wrenchValues.at(i) = wrench.get(i).asDouble();
+                    }
+                }
+            }
+            else {
+                yWarning() << LogPrefix << "[wrenchPort] read an invalid wrench bottle";
+            }
+
+            while (this->firstDataReceived != true)
+            {
+                this->firstDataReceived = true;
+            }
+        }
+
+        std::vector<double> getWrench() {
+            std::lock_guard<std::mutex> lck (mtx);
+            return this->wrenchValues;
+        }
+
+        bool getFirstDataFlag() {
+            return this->firstDataReceived;
+        }
+    };
+
+    wrenchPort leftHandFTPort;
+    wrenchPort rightHandFTPort;
+};
+
+// ==========================================
+// ICub implementation of ForceTorque6DSensor
+// ==========================================
+class ICub::ICubImpl::ICubForceTorque6DSensor
+        : public wearable::sensor::IForceTorque6DSensor
+{
+public:
+    ICub::ICubImpl* icubImpl = nullptr;
+
+    // ------------------------
+    // Constructor / Destructor
+    // ------------------------
+    ICubForceTorque6DSensor(
+            ICub::ICubImpl* impl,
+            const wearable::sensor::SensorName name = {},
+            const wearable::sensor::SensorStatus status = wearable::sensor::SensorStatus::WaitingForFirstRead) //Default sensor status
+            : IForceTorque6DSensor(name, status)
+            , icubImpl(impl)
+    {
+        //Set the sensor status Ok after receiving the first data on the wrench ports
+        if (icubImpl->leftHandFTPort.getFirstDataFlag() || icubImpl->rightHandFTPort.getFirstDataFlag()) {
+            auto nonConstThis = const_cast<ICubForceTorque6DSensor*>(this);
+            nonConstThis->setStatus(wearable::sensor::SensorStatus::Ok);
+        }
+
+    }
+
+    ~ICubForceTorque6DSensor() override = default;
+
+     void setStatus(const wearable::sensor::SensorStatus status) { m_status = status; }
+
+    // ==============================
+    // IForceTorque6DSensor interface
+    // ==============================
+    bool getForceTorque6D(Vector3& force3D, Vector3& torque3D) const override {
+        if (!icubImpl) {
```

Comment:
This should never happen. If you want to keep this check, I would put it into an `assert`.